### DISCUSSION TITLE
I see a need for stacked connector support—do you agree?

### DISCUSSION THREAD

**Discussion by gwr-de** [OpenModelica/OpenModelica#12088] [Q&A]
I would like to continue what came up in #12073 here as a separate topic. I learned that Dymola already supports use of something like a `stacked connector` (see [here](https://github.com/modelica/ModelicaSpecification/issues/3206)). For completeness, I will introduce what I mean once more.

### The Modeling Problem

As developer of the free [Business Simulation Library (BSL)](https://bsl-support.de/BusinessSimulation/BusinessSimulation), which is currently being adapted as to work more smoothly with Open Modelica, I am using the system dynamics approach in combination with Modelica's acausal connectors for a more modern take on implementing SD in Modelica.

For pre-built components in the more abstract modeling domains of business, economics, and ecology, the number of inputs and parameters will tend to grow rather quickly. You can see this in the following classes [`BusinessSimulation.Flows.Interaction.ComplexInteraction`](https://bsl-support.de/BusinessSimulation/BusinessSimulation.Flows.Interaction.ComplexInteraction.html), where `expandable connector` classes are used as outer and inner connectors.

There are couple of solutions to allow a multitude of causal (and also acausal) connectors, without making models look like "integrated circuit boards" such as `SystemDynamics.WorldDynamics.World3.Scenario_1` below with its "porcupine-like" components:

![image](https://github.com/OpenModelica/OpenModelica/assets/17766670/83122e7b-5733-4237-8986-1eed8d526a1a)

#### Arrays or Enumerations of Connectors

Arrays of connectors become inconvenient quickly if their elements are inhomogenous. What is `u[4]` again? Enumerations help, but also become unwieldy quickly imo when you have nested levels etc.

#### Composite Connectors

Nested connectors are an elegant option, as you can quickly connect lots of variables. But they are not so great, if subsets need to be connected to different elements. Modelicas connector semantics ([Section 9.1](https://specification.modelica.org/maint/3.6/connectors-and-connections.html) ff.) prevent that in some scope we can connect `a.y1` with `b.cc.u4` directly. We will need a "switch", i.e., a copy of `cc` at the respective scope. Then we can do

```
connect(a.y1., cc.u4`);
connect( cc, b.cc);
```

Should there be `input` and `ouput` prefixes then a simple copy might not suffice. So it is becoming unwieldy again imo. What if there is `b.cc1` and `d.cc2` with `cc1 =!= cc2`? This ends up in switch-board like solutions.

#### Expandable Connectors

The `expandable connector` class is a "gods end" for building more complex models and _solves_ the problem for connecting pre-built components within subsystems of nested systems in larger models such as `World3` above. Instead of connecting `n x (n-1)` components in the worst case, we will only need `n` connections to something like a "data bus."

But there are downsides for `expandable connectors` unfortunately: It is more flexible than the `composite connector` with regard to `input` and `output` prefixes and one may essentially have a global "super data bus" that is extended by all components needing input as the connector allows to have a subset of predefined variables (potentially present) to be connected. Great for testing and quick model building. But it also does not allow the following:

```
connect(a.y1, b.ec.u4); // illegal as the connected class is too deeply nested
```

Again, we will need to have a copy of the connector or simply another `expandable connector` at the current scope as described for `cc` above. This becomes unwieldy for pre-built components, where we do not want flexible, just complex connectors presented in a neat way to avoid "porcupines"?

### The Stacked Connector Approach

Why not have a bunch of unrelated connectors, some of which may even be arrays or enumerations or composite connectors ("complex plugs"), whose icons are "stacked" on top of each other? This will obviously result in completely legal Modelica code as each connector is a valid instantion on its own—it just happens to be indistinguishable from a single connector instance in Icon or Diagram View.

Enter the front end (GUI): It would be completely up to front-end-support to recognize the overlap within a "hot zone" (cf. Dag Brück, Dymola) and then assist the individual connection in a drop-down menu or similar. We would avoid all the complexities and limitations for the above stated solutions for pre-built components without long discussions in the language group.

Potentially, one may think about an `annotation` that allows to name groups of stacked connectors as to make recognition easier for tools? BTW in my library I am using a "yellow disk" within  `RealInput` and `RealOutput` connector icons as to signal that the connector is in fact an `expandable connector`—it might be the way to indicate stacked connectors in future implementations.

**What is your take on this?**

**Comment by casella** [comment]
Extending the semantics of Modelica connections to allow `connect(m1.c1.c2, m2.c3)` would open a wasp's nest of complications regarding balanced model assessments, so I would not even consider starting a discussion about that.

Dymola's stacked connector idea seems good to me, as it is very flexible and completely offloads the thing to the GUI, which avoids endless discussions by MAP-Lang.  At the source code level, you still have many individual connect statements and that ensures maximum flexibility. Supporting this feature seems to me not a big deal, and I am completely in favour of it, _provided that this possibility is explicitly mentioned by the language specification_.

What I mean is, if people start developing libraries where connectors are deliberately placed on top of each other, they won't be usable by any tool that does not consider this pattern explicitly, which would make them effectively tool-dependent.

That said, I'm not sure an annotation would be necessary, implementing a "hot zone" approach seems very straightforward, and avoid problems if you forget to include one of the connectors in the annotation.

What is still not clear to me is how you can handle the editing of the icon layer: if there are 3 stacked connectors, how can you
- select the one in the middle (e.g. to remove it)
- select the one at the bottom (e.g. to move it somewhere else)
- even know that there are other two stacked below the first one?

I guess this can be managed via right-click menus and pop-ups, but I'm not really sure.

@adrpo, @adeas31, @perost, what do you think?

**Comment by AustraleanPerson** [comment]
@gwr69 can you explain how exactly `BusinessSimulation.Flows.Interaction.ComplexInteraction` can benefit from stacked connectors?

**Reply by gwr-de** [reply]
Sure. I would like to point you at `Examples.SimpleProductionChainIII`:
![](https://bsl-support.de/BusinessSimulation/resources/diagrams/BusinessSimulation.Examples.SimpleProductionChainIII_diagram_large.png)

There we have a similar class [`Diffusion`](https://bsl-support.de/BusinessSimulation/BusinessSimulation.MoleculesOfStructure.Actuators.Diffusion.html) that uses `expandable connector` classes for inputs and outputs. In such a model, I would believe most people would like to click a representative connector icon to then have a dropdown menu select a relevant connector that is then connected to another input connector directly. Unfortunately, the expandable connector—at least in System Modeler—supports drop-down menu access to a subconnector, but we need to add another `expandable connector` at the level of the model to mediate the connection. That is cumbersome for components that do not need their connectors to be “expandable” but simply to act as a “collector” (rather than a combinator.

I believe stacked connector semantics would be rather helpful in such cases without adding anything new to Modelica specs. Do you follow?

**Comment by AustraleanPerson** [comment]
> I would believe most people would like to click a representative connector icon to then have a dropdown menu

In the example above, there is no dropdown menu when connecting because it is not a DataOutPort instance.
But the tool can help with this. All you have to do is start making a connection and select 'create a node'.
Since the buses are marked with a yellow circle, the user will always know what to do. Its redundant but easy step.
If the library design assumes that the buses will always face up, the connector icon can be made to face up by default. Then the user will not even have to turn it over.
<kbd> ![1](https://github.com/OpenModelica/OpenModelica/assets/147278612/9c4dd378-bbfd-449e-806d-dd3d244b5688)  </kbd>

<kbd>![2](https://github.com/OpenModelica/OpenModelica/assets/147278612/7fea70cc-6359-4475-b18b-967e71f640bf) </kbd>

**Reply by gwr-de** [reply]
This "creation of a node" is also supported by WSM and thus I am familiar with it. Note that—given input and output prefixes for _composite connectors_ —this solution works nicely only for _expandable connector_ classes imo.

**Reply by AustraleanPerson** [reply]
Composite connectors (not expandable) can be used. Why not? But in that case Rate should be a connector
`connector RateOutput = output Rate;`

**Comment by AustraleanPerson** [comment]
>There are couple of solutions to allow a multitude of causal (and also acausal) connectors, without making models look like "integrated circuit boards"

If there are too many lines, they can become confusing. You can use this library component design:
![4](https://github.com/OpenModelica/OpenModelica/assets/147278612/355d447a-1558-4c47-9d0e-c6fd4625760d)

Autocompletion will certainly help.

![3](https://github.com/OpenModelica/OpenModelica/assets/147278612/b58445db-8b57-4d24-b1e2-1afe1669aba2)

>That is cumbersome for components that do not need their connectors to be “expandable”

In that approach they can be just 'plain' connectors.

**Reply by gwr-de** [reply]
Is what you are proposing the following:
1. Make `dataOut` for `wordOfMouth` a `connector` (i.e., a composite one) instead of an expandable connector.
2. Avoiding `connect-equation` semantics by directly binding `totalOrders.u2` to an expression, e.g., `wordOfMouth.dataOut.totalAdoptionRate`?
3. Forfeiting on showing any graphical representation for the binding, e.g., no line connection?

**Reply by casella** [reply]
3. would be quite confusing IMHO

**Reply by gwr-de** [reply]
@casella I totally agree, which is why—if my understanding of the proposal made by @AustraleanPerson is correct—I would see this as an unfavourable solution best avoided and left for "lack of better solutions."
